Feature/#297 mimic cxr dataset

marlenesteiner opened the pull request:
I will add some more comments to the jupiter notebook, this is still WIP. So far, I only got the X and T matrix, since I need some advice for the Z matrix. 

Linked issue: #297

anasedova commented on examples/data_preprocessing/mimic_dataset/prepare_mimic_cxr.ipynb:
```diff
@@ -0,0 +1,1520 @@
+   "metadata": {},
+   "source": [
+    "To get one label for each study, we do the following:\n",
+    "If there is exactly one label positively mentioned, this label will be assigned to the study. If there are multiple labels positively mentioned, one label will be chosen randomly. When there is no 1.0 assigned to the study, the label will be set to 'No Finding'."
```
if you are transforming the data to Knodle input, why would you need labels?

marlenesteiner replied:
I use the labels for the image encoding (second approach including finetuning).

anasedova commented on examples/data_preprocessing/mimic_dataset/prepare_mimic_cxr.ipynb:
```diff
@@ -0,0 +1,1520 @@
+   ],
+   "source": [
+    "#change names to ids\n",
+    "labels = {'Atelectasis':0,\n",
```
```
labels = {cat: id for (cat, id) in enumerate(labels_chexpert.columns)}
```

anasedova commented on examples/data_preprocessing/mimic_dataset/prepare_mimic_cxr.ipynb:
```diff
@@ -0,0 +1,1520 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Next, the records, reports, labels and split are merged to one file. Note that sometimes one report refers to multiple X-rays (e.g. one frontal and one lateral X-ray) and since the labels are derived from the reports, one label might refer to multiple images."
```
Here, again, there shouldn't be any labels

anasedova commented on examples/data_preprocessing/mimic_dataset/prepare_mimic_cxr.ipynb:
```diff
@@ -0,0 +1,1520 @@
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "train-val-test split proportions: split\n",
```
statistics is good, but I think it also should be moved from there to mb the part where you download the reports

anasedova commented on examples/data_preprocessing/mimic_dataset/prepare_mimic_cxr.ipynb:
```diff
@@ -0,0 +1,1520 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "We are using the [\"mentions\" from Chexpert-labler](https://github.com/stanfordmlgroup/chexpert-labeler/tree/master/phrases/mention) for building our rules. Therefore, we need to download the .txt files corresponding to each class from github. The file names are the labels we defined above, where all letters are lowercase and instead of whitespaces there are underscores. So first we do some ...
```
```suggestion
    "We are using the [\"mentions\" from Chexpert-labler](https://github.com/stanfordmlgroup/chexpert-labeler/tree/master/phrases/mention) for building our rules. Therefore, we need to download the TXT files corresponding to each class from GitHub. The file names are the labels we defined above, where all letters are lowercase and instead of whitespaces, there are underscores. So first we do some transformations to get a list of our classes. "
```

+ mention somehow that our rules are keywords
+ mention that you rules are for annotation of the reports in order to *later* use them to calculate labels, which, in turn, can be later used to annotate images. Now this whole process and the reasons why you need rules is not clear.

marlenesteiner replied:
Is this now clear with the description and graph at the beginning?

anasedova commented on examples/data_preprocessing/mimic_dataset/prepare_mimic_cxr.ipynb:
```diff
@@ -0,0 +1,1520 @@
+    "classes = [each_string.lower() for each_string in classes]\n",
+    "#replace whitespace with _\n",
+    "classes = [each_string.replace(\" \", \"_\") for each_string in classes]\n",
+    "labels2ids = {classes[i]:i for i in range(14)}"
```
what is the difference between `labels2ids` and `labels`?

marlenesteiner replied:
In labels the classes start with capital letters and include whitespaces (like the colnames from the "mimic-cxr-2.0.0-chexpert.csv") and in labels2ids the labels are all lowercase and no whitespace, (like the filenames from Chexpert mentions).  Is there a better way to deal with those two spelling variants than two dictionaries?

anasedova commented on examples/data_preprocessing/mimic_dataset/prepare_mimic_cxr.ipynb:
```diff
@@ -0,0 +1,1520 @@
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "#create folder chexpert_rules\n",
```
I would switch this cell and the previous one

marlenesteiner replied:
But I need the "classes" which are generated in the previous cell for this one.

anasedova commented on examples/data_preprocessing/mimic_dataset/prepare_mimic_cxr.ipynb:
```diff
@@ -0,0 +1,1520 @@
+    }
+   ],
+   "source": [
+    "#building the dataframe \"rules\"\n",
```
what is the dataframe "rules"?

anasedova commented on examples/data_preprocessing/mimic_dataset/prepare_mimic_cxr.ipynb:
```diff
@@ -0,0 +1,1520 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Now we want to check if there are any rules which assign to the same class. Indeed, there is one case: rule \"defib\" builds a rule for two different classes. In our T matrix, \"defib\" occures in two rows, since knodle cannot handle one rule assigned to multiple classes. "
```
Let's have it for now not as separate T rows, but as one rule related to two classes at the same time. It would work in the next Knodle release.

marlenesteiner replied:
Ok, then I need to delete one row and change one 0 to 1 in the according column in the remaining row manually.

anasedova commented on examples/data_preprocessing/mimic_dataset/prepare_mimic_cxr.ipynb:
```diff
@@ -0,0 +1,1520 @@
+   "metadata": {},
+   "source": [
+    "### Z Matrix\n",
+    "The Z matrix contains the information wheter a rule matches to an observation or not. Its shape is #instances x #rules. "
```
```suggestion
    "The Z matrix contains the information about the rules matched to the reports. Its shape is #instances x #rules. "
```

I would avoid "observation" term in order to avoid confusion and use words "rule" or "keywords" instead.

anasedova commented on examples/data_preprocessing/mimic_dataset/prepare_mimic_cxr.ipynb:
```diff
@@ -0,0 +1,1520 @@
+    "print(len(np.unique(rules['rule'])) == len(rules['rule']))\n",
+    "rules_size = rules.groupby('rule').size()\n",
+    "print(rules_size[np.where(rules_size > 1)[0]])\n",
+    "#rule defib appears for two different classes"
```
dump the ready T matrix?

anasedova commented on examples/data_preprocessing/mimic_dataset/prepare_mimic_cxr.ipynb:
```diff
@@ -0,0 +1,1520 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "In the second approach of image encoding, we use the concept of transfer learning. \n",
```
I missed the first one :D

marlenesteiner replied:
Ah ok, I added 'first approach' and 'second approach' in the headings, so it should be clearer now. Basically in the first approach I don't do any finetuning and in the second I finetune the model with the Chexpert labels. 

The reason for the two approaches is: The second approach uses the Chexpert labels for finetuning and I was not sure if we want to touch our labels there since this might be confusing with our goal to demonstrate weak supervision. 
I think it depends whether we evaluate our model with gold labels in the end or (if we don't find any appropriate gold labels) we have to stick with the Chexpert labels.  In the latter case, I would not use the Chexpert labels for the image encoding, but rather go with the first approach (without finetuning).
Alternatively I could finetune the model for the image encoding on different data, like  iuxray.

marlenesteiner replied:
Since we have a test set from the chexpert data, the finetuning with the weak labels sounds fine for me. But I think I will leave both approaches and compare the results in the end.

anasedova commented on examples/data_preprocessing/mimic_dataset/prepare_mimic_cxr.ipynb:
```diff
@@ -0,0 +1,1520 @@
+   "source": [
+    "In the second approach of image encoding, we use the concept of transfer learning. \n",
+    "Therefore, we take a pretrained CNN and continue training the model with our data. This saves a lot of time and leads to satisfying results even on a small dataset. \n",
+    "For the implementation we again use pytorch and the pretrained resnet50 from torchvision. "
```
```suggestion
    "For the implementation we use PyTorch library and the pretrained ResNet50 from `torchvision` library. "
```

+ add link to ResNet40

marlenesteiner replied:
Is it enough to link to ResNet the first time I mention it, or should I link it every time?

anasedova commented on examples/data_preprocessing/mimic_dataset/prepare_mimic_cxr.ipynb:
```diff
@@ -0,0 +1,1520 @@
+   "metadata": {},
+   "source": [
+    "In the second approach of image encoding, we use the concept of transfer learning. \n",
+    "Therefore, we take a pretrained CNN and continue training the model with our data. This saves a lot of time and leads to satisfying results even on a small dataset. \n",
```
```suggestion
    "Therefore, we take a pretrained CNN and continue training the model with our data. This saves a lot of time and leads to satisfying results. \n",
```

anasedova commented on examples/data_preprocessing/mimic_dataset/prepare_mimic_cxr.ipynb:
```diff
@@ -0,0 +1,1520 @@
+   "outputs": [],
+   "source": [
+    "class_counts = np.zeros(14)\n",
+    "for i in range(14): class_counts[i] = sum(input_train[:,6]==i)\n",
```
```suggestion
    "for i in range(num_classes):"
    "     class_counts[i] = sum(input_train[:,6]==i)\n",
```

I remember there is no `num_classes`parameter defined, but I strongly recommend you to define it :) usually it's better not to use any hardcode parameters (e.g. `14`), but to introduce them once in the beginning and use the variables instead

anasedova commented on examples/data_preprocessing/mimic_dataset/prepare_mimic_cxr.ipynb:
```diff
@@ -0,0 +1,1520 @@
+   "metadata": {},
+   "source": [
+    "Now we initialize the model:\n",
+    "We change the last layer to 14, since we only have 14 classes. We choose the Cross Entropy Loss as an optimization cirterion, the adam optimizer and an adaptive learning rate. \n",
```
```suggestion
    "The size of the output layer equals the number of classes (14 classes in our case). We choose the CrossEntropy as a loss function, the Adam optimizer and an adaptive learning rate. \n",
```

anasedova commented on examples/data_preprocessing/mimic_dataset/prepare_mimic_cxr.ipynb:
```diff
@@ -0,0 +1,1520 @@
+    }
+   ],
+   "source": [
+    "#save feature matrix\n",
```
this comment could be a markdown text and then the `dump` code to the separate cell in order to highlight it

(also: mb I just overlooked it, but - have you saved the reports as well?)

marlenesteiner replied:
I have the reports in one column of the 'input_list_full' but no file with only the reports.

anasedova commented on examples/data_preprocessing/mimic_dataset/prepare_mimic_cxr.ipynb:
```diff
@@ -0,0 +1,1520 @@
+ },
+ "nbformat": 4,
+ "nbformat_minor": 4
+}
```
some final word is needed

anasedova commented on examples/data_preprocessing/mimic_dataset/prepare_mimic_cxr.ipynb:
```diff
@@ -0,0 +1,1520 @@
+   ],
+   "source": [
+    "#save the whole file\n",
+    "dump(input_list_full, \"input_list.lib\")"
```
the names if the output files whould be defined in the very beginning as well

marlenesteiner replied:
Where would be a good place for that? Before or after the library imports?

anasedova reviewed (changes requested):
Awesome, you've done a lot of work! Here are a couple of my suggestions to the notebook - pls have a look at them. I will review the .py file more precisely in the next iteration review. 

I would also ask you to mark the conversation as resolved when you made the correction and don't have any questions or add the comments in the corresponding conversation if smth is not clear/you do not agree with my suggestion/you have another idea - that would make it much easier for me to check everything afterward :)

marlenesteiner commented on examples/data_preprocessing/mimic_dataset/prepare_mimic_cxr.ipynb:
```diff
@@ -0,0 +1,1520 @@
+   "metadata": {},
+   "source": [
+    "To download the images with the following code, a credentialed PhysioNet account is required. Before executing the following you should do the following:\n",
+    "1. check if you have enough memory space in the current environment (the first 1000 images are 1.51GB large)\n",
```
I meant the current directory, since the images will be downoaded there. I changed it to "check if you have enough memory space in the current directory"

marlenesteiner commented on examples/data_preprocessing/mimic_dataset/prepare_mimic_cxr.ipynb:
```diff
@@ -0,0 +1,1520 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Downloading MIMIC-CXR-JPG images"
```
In the "getting all set up" only the lists are downloaded which contain the paths of each image and each report. These paths I use for downloading the images and reports seperatly. I will explain this in the next version!

marlenesteiner replied:
In the new version, no files need to be downloaded in the "getting all set up", but all files will be automatically downloaded in the "downloads" section. You are right, this is more intuitive!!

marlenesteiner commented on examples/data_preprocessing/mimic_dataset/prepare_mimic_cxr.ipynb:
```diff
@@ -0,0 +1,1520 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Some reports refer to multiple images. Now we merge the reports to our record list such that each image gets a report assigned if a report is available. "
```
Is the following clearer?
"Now we merge the reports to our record list such that each image gets a report assigned if a report is available. 
Some reports refer to multiple images. In this case the report is merged to all refered images."

marlenesteiner commented on examples/data_preprocessing/mimic_dataset/prepare_mimic_cxr.ipynb:
```diff
@@ -0,0 +1,1520 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Labels"
```
Do you mean the whole labels section (including code) or just the description of the CheXpert output?

anasedova reviewed (changes requested):
Great, it is almost done! Just a couple of tiny corrections. The only thing that isn't quite clear to me is about the T matrix: what did you mean by manual deletion of some rows? 

(also: did you inherit the approach of working with Impression section, or with Findings if Impression is not available, from the original paper? )

After you make these corrections, please let me know. The next step would be a second review from @AndSt, and we could merge it afterward.

marlenesteiner commented:
Thank you for reviewing! Now I have incorporated all your comments. 

I changed the structure, such that the labels are only included in the last section before finetuning, now it is more clear that labels are not used in any other data preprocessing step. 

To your question ("what did you mean by manual deletion of some rows?"): There is one rule referring to two classes. At the moment, there are two rows for this rule in the T matix, because the code automatically produces two rows. If there should be only one row with two "1"s, I would have to change it manually, so delete one row and insert a "1" in the other. Should I do that, or should I just leave the two rows? 

I also tried the efficientnet Vasiliki suggested, but the results from resnet50 were better, so I would stick to the resnet50.

anasedova commented:
> To your question ("what did you mean by manual deletion of some rows?"): There is one rule referring to two classes. At the moment, there are two rows for this rule in the T matix, because the code automatically produces two rows. If there should be only one row with two "1"s, I would have to change it manually, so delete one row and insert a "1" in the other. Should I do that, or should I just leave the two rows?

Hmm nice question. I think two rows isn't a good option, since then the Z and T matrices would have different dimensions (and, as a result, we won't be able to multiply them in order to get the labels). The scenario with two "1"s in a row is not currently supported in Knodle (but is in development and will be supported in the nearest future). 
My suggestion would be to have both options saved: (1) choose randomly one class and have T of one-hot-vectors only, (2) T with one row where there are two "1"s. 

(quite complicated handling of one single rule though 🤨 but still, if we don't do that, we could have problems with multiplying matrices and launching trainers)

anasedova commented:
@vasilikikou could you have a look at this PR as well? Thank you!

marlenesteiner commented:
> 
Actually Z and T matrices still have the same dimension, since the rule appears twice in both matrices. 
So the dimension of my z matrix is (964, 148)
and the dimension of my t matrix is (148, 14)

vasilikikou commented on examples/data_preprocessing/mimic_dataset/prepare_mimic_cxr.ipynb:
```diff
@@ -0,0 +1,1862 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Now we want to check if there are any rules which assign to the same class. Indeed, there is one case: rule \"defib\" builds a rule for two different classes. In our T matrix, \"defib\" occures in two rows. "
```
```suggestion
    "Now we want to check if there are any rules, which assigns 1 to the same class. Indeed, there is one case: rule \"defib\" builds a rule for two different classes. In our T matrix, \"defib\" occures in two rows. "
```

vasilikikou commented on examples/data_preprocessing/mimic_dataset/prepare_mimic_cxr.ipynb:
```diff
@@ -0,0 +1,1862 @@
+   "source": [
+    "# Image encoding \n",
+    "For the image encoding we try two different approaches:\n",
+    "1. extract features using a pretrained model (no labels)\n",
```
```suggestion
    "1. extract features using a model pretrained on ImageNet (no finetuning)\n",
```

vasilikikou commented on examples/data_preprocessing/mimic_dataset/prepare_mimic_cxr.ipynb:
```diff
@@ -0,0 +1,1862 @@
+    "# Image encoding \n",
+    "For the image encoding we try two different approaches:\n",
+    "1. extract features using a pretrained model (no labels)\n",
+    "2. finetune a pretrained model and then extract features (finetuning with weak chexpert labels)\n",
```
```suggestion
    "2. finetune a model pretrained on ImageNet and then extract features (finetuning with weak chexpert labels)\n",
```

vasilikikou commented on examples/data_preprocessing/mimic_dataset/prepare_mimic_cxr.ipynb:
```diff
@@ -0,0 +1,1862 @@
+    "1. extract features using a pretrained model (no labels)\n",
+    "2. finetune a pretrained model and then extract features (finetuning with weak chexpert labels)\n",
+    "\n",
+    "## First approach: Extracting features using a pretrained model (no labels)"
```
```suggestion
    "## First approach: Extracting features using a pretrained model (no finetuning)"
```

vasilikikou commented on examples/data_preprocessing/mimic_dataset/prepare_mimic_cxr.ipynb:
```diff
@@ -0,0 +1,1862 @@
+   "metadata": {},
+   "source": [
+    "For the implementation we use PyTorch and the pretrained ResNet50 from `torchvision` library. \n",
+    "The following class loads the data and transforms it in the way it is required for [ResNet50](https://arxiv.org/abs/1512.03385). It is written in the form such that it is compatible with `torch.utils.data.DataLoader`. "
```
```suggestion
    "The following class loads the data and transforms it in the way it is required for [ResNet50](https://arxiv.org/abs/1512.03385). It is written in a form such that it is compatible with `torch.utils.data.DataLoader`. "
```

vasilikikou commented on examples/data_preprocessing/mimic_dataset/prepare_mimic_cxr.ipynb:
```diff
@@ -0,0 +1,1862 @@
+    "    def __getitem__(self, index):\n",
+    "        'Generates one sample of data'\n",
+    "        # Select sample\n",
+    "        image = Image.open(self.path[index,3].replace(\".dcm\", \".jpg\")).convert('RGB')\n",
```
Choose if you will use single or double quotes and keep the same throughout the code.

vasilikikou commented on examples/data_preprocessing/mimic_dataset/prepare_mimic_cxr.ipynb:
```diff
@@ -0,0 +1,1862 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "We load the ResNet50 model and remove the last layer, so the output is then the second last layer with dimension 1x2048. These are our features. "
```
```suggestion
    "We load the ResNet50 model and remove the last layer, so the output is then extracted from the second last layer with dimension 1x2048. These are our features. "
```

vasilikikou commented on examples/data_preprocessing/mimic_dataset/prepare_mimic_cxr.ipynb:
```diff
@@ -0,0 +1,1862 @@
+   "metadata": {},
+   "source": [
+    "In the second approach of image encoding, we use the concept of transfer learning. \n",
+    "Therefore, we take a pretrained CNN and continue training the model with our data. This saves a lot of time and often leads to satisfying results. \n",
```
```suggestion
    "Therefore, we take a CNN pretrained on ImageNet and continue training the model with our data. This saves a lot of time and often leads to satisfying results. \n",
```

vasilikikou replied:
Also I don't understand this sentence: "This saves a lot of time and often leads to satisfying results.". Since the finetuning takes some time, this approach doesn't save time compared to the first approach. Moreover, it's important to mention that images from ImageNet are very different are different from medical images, so this finetuning can help improve the results.

vasilikikou commented on examples/data_preprocessing/mimic_dataset/prepare_mimic_cxr.ipynb:
```diff
@@ -0,0 +1,1862 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "The CheXpert labler detects the presence of 14 diagnoses in radiology reports. \n",
```
```suggestion
    "The CheXpert labeler detects the presence of 14 diagnoses in radiology reports. \n",
```

vasilikikou commented on examples/data_preprocessing/mimic_dataset/prepare_mimic_cxr.ipynb:
```diff
@@ -0,0 +1,1862 @@
+    "* -1.0 The label was either: (1) mentioned with uncertainty in the report, and therefore may or may not be present to some degree in the corresponding image, or (2) mentioned with ambiguous language in the report and it is unclear if the pathology exists or not\n",
+    "* Missing (empty element) - No mention of the label was made in the report\n",
+    "\n",
+    "(Johnson, Lungren et al. (2019)\n",
```
```suggestion
    "(Johnson, Lungren et al. (2019))\n",
```

vasilikikou commented on examples/data_preprocessing/mimic_dataset/prepare_mimic_cxr.ipynb:
```diff
@@ -0,0 +1,1862 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "First, we set a number of images which should be used for finetuning. If the number of images downloaded is less than 750, all images are used, otherwise we will use 750."
```
```suggestion
    "First, we set a number of images, which should be used for finetuning. If the number of images downloaded is less than 750, all images are used, otherwise we will use 750."
```

vasilikikou commented on examples/data_preprocessing/mimic_dataset/prepare_mimic_cxr.ipynb:
```diff
@@ -0,0 +1,1862 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Then we split the finetuning data randomly into 80% traininag and 20% validation data."
```
```suggestion
    "Then we split the finetuning data randomly into 80% training and 20% validation data."
```

vasilikikou commented on examples/data_preprocessing/mimic_dataset/prepare_mimic_cxr.ipynb:
```diff
@@ -0,0 +1,1862 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "The following function which trains the model is taken from https://pytorch.org/tutorials/beginner/transfer_learning_tutorial.html"
```
```suggestion
    "The following function, which trains the model is taken from https://pytorch.org/tutorials/beginner/transfer_learning_tutorial.html"
```

vasilikikou commented on examples/data_preprocessing/mimic_dataset/prepare_mimic_cxr.ipynb:
```diff
@@ -0,0 +1,1862 @@
+   "source": [
+    "As test data we use the CheXpert validation data, which contains gold labels (Irvin, Rajpurkar et al. 2019). \n",
+    "\n",
+    "The data can be found on [this homepage](https://stanfordmlgroup.github.io/competitions/chexpert/). After the use agreement was signed, you recieve an e-mail with two links. One for the dataset at high-resolution and one at downsampled resolution. Here we use the latter. \n",
```
```suggestion
    "The data can be found on [this homepage](https://stanfordmlgroup.github.io/competitions/chexpert/). After the use agreement is signed, you receive an e-mail with two links. One for the dataset at high-resolution and one at downsampled resolution. Here we use the latter. \n",
```

vasilikikou reviewed (changes requested):
Looks great! I added comments suggesting minor changes.

vasilikikou reviewed (approved):
Great work Marlene! Thank you for solving all the comments! I think it's ready to merge :)
